src/pam-ssh-add/pam-ssh-add.c

```diff
@@ -455,8 +456,9 @@ setup_child (const char **args,
   if (setegid (getgid ()) < 0 || seteuid (getuid ()) < 0)
       error ("failed to restore credentials");
 
-  /* Setup process credentials */
-  if (setgid (pwd->pw_gid) < 0 || setuid (pwd->pw_uid) < 0 ||
+  /* Setup process credentials; if we have root privs, drop any auxiliary groups too */
+  if ((geteuid() == 0 ? initgroups(pwd->pw_name, pwd->pw_gid) < 0 : 0) ||
```

Comment:
I think that having the uid 0 doesn't always mean you have permission to change the group. Can we base the check on the intent instead? if getgid() != pwd->pw_gid then we call initgroups? Does that work?

Reply:
Ah, nice idea! Yes, that is better, I updated the commit. (I used `getegid()` though, I think that's slightly more correct).

Reply:
We explicitly setegid with getgid above so by this point they should always be the same. But yeah this seems good.